### DISCUSSION TITLE
Setting the pcam's cullingmask layers help

### DISCUSSION THREAD

**Discussion by Zoo4125** [ramokz/phantom-camera#202] [Question / Help Wanted]
Hi,
I was trying to turn the cullmask layers 20 & 17 off on the MainCamera
I then tried doing this through script in the player_controller_third_person.gd from the 3DFollowThirdPersonExampleScene which I have built off of, so the scene is set up pretty much identical to that. I set the fov here and that worked fine for the MainCamera3D but I couldn't get the cullmask layers, I wanted to turn off am I trying in the right place? and maybe could you provide an example in code of how I could turn off specific layers. Thank you very much in advance. Love the add-on
![Screenshot_6](https://github.com/ramokz/phantom-camera/assets/46336071/70fc0e91-0811-4cf5-8136-140ff3d13e9e)

**Comment by ramokz** [comment]
So setting a `cull_mask` is a bit tricky at first, but fairly straightforward once you know.

Essentially, the `cull_mask` is a sum of a bit value, where each `cull_mask` layer represents a part of that combined value. E.g. layer 1 = 1bit, layer 2 = 2bit, layer 3 = 4bit, layer 4 = 8bit etc.
[Godot has a documentation page](https://docs.godotengine.org/en/stable/tutorials/physics/physics_introduction.html#code-example) with better examples than above.

So at the minute, you would have to calculate the value and then add or subtract the layer you want to include or exclude respectively from the `cull_mask`. Can see how it's done [internally in CPP in the source code](https://github.com/godotengine/godot/blob/51991e20143a39e9ef0107163eaf283ca0a761ea/scene/3d/camera_3d.cpp#L638).

What I can do, though, is add a similar helper function to the one in the `Camera3D` class has, so you would effectively only need to write one line to enable or disable a layer, like so:
```gdscript
pcam.set_camera_cull_mask(LAYER_NUMBER, true/false)
```

**Reply by Zoo4125** [reply]
I was looking into it, and came up with something involving the solutions the documentation I could find suggested. But I couldn't get it to work unfortunately. I'll probably take another crack at it with the additional resources you gave me, thanks very much.

I do think that having a helper function like the camera has sounds like a good feature for phantom camera to also have and would really simplify toggling the culling mask and would be appreciated.

**Reply by ramokz** [reply]
Got the helper function working, so should have a PR merged into main within long. (Issue #204)
I'll drop a copy of the modified `PhantomCamera3D.gd` into this discussion post so you don't have to wait for the next release.

**Reply by ramokz** [reply]
Have merged in #205. This won't be updated in the addon right away, so for a quick update, you can replace your current `phantom_camera_3D.gd` with the one attached below. 
[phantom_camera_3D.gd.zip](https://github.com/ramokz/phantom-camera/files/14115495/phantom_camera_3D.gd.zip)

After that, you can enable or disable individual `cull mask` layers using:
```gdscript
pcam.set_cull_mask_value(layer_number, true/false)
```
